File: proxy_switch/features/docker.py
"""Docker daemon proxy and registry mirror configuration.

Proxy: systemd drop-in at /etc/systemd/system/docker.service.d/proxy.conf
       (daemon.json cannot set environment variables for dockerd).
Mirror: registry-mirrors in /etc/docker/daemon.json.
"""

from __future__ import annotations

import json
from typing import Dict, List

from ..core.models import Result, StatusInfo

NAME = "docker"
DESCRIPTION = "Docker daemon proxy and registry mirror"
CONFIG_FILES = [
    "/etc/systemd/system/docker.service.d/proxy.conf",
    "/etc/docker/daemon.json",
]

DOCKER_CONF_DIR = "/etc/systemd/system/docker.service.d"
DOCKER_CONF = "/etc/systemd/system/docker.service.d/proxy.conf"
DAEMON_JSON = "/etc/docker/daemon.json"
SUPPORTS_MIRROR = True


def _read_file(executor, path: str) -> str:
    if executor:
        r = executor.run(f"cat {path} 2>/dev/null || echo ''")
        return r.stdout
    try:
        with open(path) as f:
            return f.read()
    except (FileNotFoundError, PermissionError):
        return ""


def _write_file(executor, path: str, content: str, sudo: bool) -> bool:
    if executor:
        r = executor.write(path, content, sudo=sudo)
        return r.returncode == 0
    import tempfile, subprocess, os
    try:
        with tempfile.NamedTemporaryFile(mode="w", delete=False) as tf:
            tf.write(content)
            tf.flush()
        if sudo:
            r = subprocess.run(["sudo", "cp", tf.name, path],
                               capture_output=True, text=True)
        else:
            import shutil
            shutil.copy(tf.name, path)
            r = subprocess.CompletedProcess([], 0)
        os.unlink(tf.name)
        return r.returncode == 0
    except Exception:
        return False


def _proxy_content(proxy_config: Dict[str, str]) -> str:
    http = proxy_config.get("http_proxy", "")
    https = proxy_config.get("https_proxy", "")
    no_proxy = proxy_config.get("no_proxy", "")
    lines = ["# Managed by proxy-switch", "[Service]"]
    if http:
        lines.append(f'Environment="HTTP_PROXY={http}"')
        lines.append(f'Environment="http_proxy={http}"')
    if https:
        lines.append(f'Environment="HTTPS_PROXY={https}"')
        lines.append(f'Environment="https_proxy={https}"')
    if no_proxy:
        lines.append(f'Environment="NO_PROXY={no_proxy}"')
        lines.append(f'Environment="no_proxy={no_proxy}"')
    return "\n".join(lines) + "\n"


def _restart_docker(executor) -> List[str]:
    """Restart Docker daemon. Returns list of errors."""
    errs = []
    if executor:
        for cmd in ["sudo systemctl daemon-reload", "sudo systemctl restart docker"]:
            r = executor.run(cmd)
            if r.returncode != 0:
                errs.append(f"{cmd}: {r.stderr}")
    else:
        import subprocess
        for cmd in ["sudo systemctl daemon-reload", "sudo systemctl restart docker"]:
            r = subprocess.run(cmd, shell=True, capture_output=True, text=True)
            if r.returncode != 0:
                errs.append(f"{cmd}: {r.stderr}")
    return errs


def _update_daemon_json(content: str, mirror: str) -> str:
    """Add or update registry-mirrors in daemon.json content."""
    try:
        cfg = json.loads(content) if content.strip() else {}
    except json.JSONDecodeError:
        cfg = {}
    if mirror:
        cfg["registry-mirrors"] = [mirror]
    elif "registry-mirrors" in cfg:
        del cfg["registry-mirrors"]
    return json.dumps(cfg, indent=2) + "\n" if cfg else ""


def detect(executor=None) -> bool:
    if executor:
        r = executor.run("command -v docker && systemctl show -p Id docker.service 2>/dev/null")
        return r.returncode == 0 and "docker.service" in r.stdout
    import shutil
    if not shutil.which("docker"):
        return False
    import subprocess
    try:
        r = subprocess.run(["systemctl", "show", "-p", "Id", "docker.service"],
                           capture_output=True, text=True, timeout=5)
        return r.returncode == 0 and "docker.service" in r.stdout
    except Exception:
        return False


def enable(proxy_config: Dict[str, str], executor=None) -> Result:
    """Set Docker proxy and optionally registry mirror, then restart."""
    # Early sudo check
    if executor and not executor.has_sudo():
        return Result(
            success=False,
            message="Sudo access required. Configure passwordless sudo on the remote server."
        )

    content = _proxy_content(proxy_config)
    ok = _write_file(executor, DOCKER_CONF, content, sudo=True)
    if not ok:
        return Result(success=False, message=f"Failed to write {DOCKER_CONF}")

    # Registry mirror
    mirror = proxy_config.get("mirror", "")
    if mirror:
        existing = _read_file(executor, DAEMON_JSON)
        updated = _update_daemon_json(existing, mirror)
        if updated:
            _write_file(executor, DAEMON_JSON, updated, sudo=True)

    errs = _restart_docker(executor)
    if errs:
        return Result(success=True, message="Proxy written, restart had issues",
                      details="; ".join(errs))
    return Result(success=True, message="Docker proxy configured and restarted")


def disable(executor=None) -> Result:
    """Remove Docker proxy config and restart daemon."""
    if executor:
        executor.run(f"sudo rm -f {DOCKER_CONF}")
    else:
        import subprocess
        subprocess.run(["sudo", "rm", "-f", DOCKER_CONF], shell=True)

    # Remove registry-mirrors from daemon.json if present
    existing = _read_file(executor, DAEMON_JSON)
    if existing.strip():
        updated = _update_daemon_json(existing, "")
        if updated != existing:
            _write_file(executor, DAEMON_JSON, updated, sudo=True)

    _restart_docker(executor)
    return Result(success=True, message="Docker proxy disabled")


def status(executor=None) -> StatusInfo:
    """Check Docker proxy and registry mirror status."""
    content = _read_file(executor, DOCKER_CONF)
    enabled = "HTTP_PROXY=" in content
    proxy = None
    for line in content.split("\n"):
        if 'HTTP_PROXY=' in line:
            import re
            m = re.search(r'"(.*?)"', line)
            if m:
                proxy = m.group(1)
                break

    mirror = None
    daemon = _read_file(executor, DAEMON_JSON)
    if daemon.strip():
        try:
            cfg = json.loads(daemon)
            mirrors = cfg.get("registry-mirrors", [])
            if mirrors:
                mirror = mirrors[0]
        except json.JSONDecodeError:
            pass

    return StatusInfo(
        enabled=enabled,
        proxy=proxy,
        mirror=mirror,
        config_file=DOCKER_CONF,
        notes="Docker daemon restarted to apply changes",
    )


def validate(executor=None) -> List[str]:
    issues = []
    if not detect(executor):
        issues.append("Docker is not installed or not managed by systemd")
    return issues

File: src-tauri/src/proxy/remote/docker.rs
use crate::models::{ComponentId, ManualStep, OpResult, ProxyConfig, ProxyStatus};
use crate::proxy::ProxyModule;
use crate::ssh::connection::SshSession;

const DOCKER_DIR: &str = "/etc/systemd/system/docker.service.d";
const DOCKER_CONF: &str = "/etc/systemd/system/docker.service.d/proxy.conf";
const DAEMON_JSON: &str = "/etc/docker/daemon.json";

pub struct DockerRemoteModule;

impl ProxyModule for DockerRemoteModule {
    fn name(&self) -> &'static str {
        "docker_remote"
    }
    fn description(&self) -> &'static str {
        "Docker daemon proxy"
    }
    fn config_files(&self) -> Vec<String> {
        vec![DOCKER_CONF.into(), DAEMON_JSON.into()]
    }
    fn manual_steps(&self) -> Vec<(String, Vec<String>)> {
        vec![(
            "Configure Docker proxy".into(),
            vec![
                format!("sudo mkdir -p {}", DOCKER_DIR),
                format!("sudo nano {}", DOCKER_CONF),
                String::new(),
                "# Add these lines:".to_string(),
                "[Service]".into(),
                "Environment=\"HTTP_PROXY=http://your-proxy:port\"".into(),
                "Environment=\"HTTPS_PROXY=http://your-proxy:port\"".into(),
                "Environment=\"NO_PROXY=localhost,127.0.0.1\"".into(),
                String::new(),
                "# Then restart:".to_string(),
                "sudo systemctl daemon-reload".into(),
                "sudo systemctl restart docker".into(),
            ],
        )]
    }
    fn detect(&self, session: &SshSession) -> bool {
        session.tool_exists("docker")
            && session
                .run("systemctl show -p Id docker.service 2>/dev/null", 5)
                .stdout
                .contains("docker.service")
    }
    fn status(&self, session: &SshSession) -> ProxyStatus {
        let content = session.read_file(DOCKER_CONF);
        let enabled = content.contains("HTTP_PROXY=");
        let mut proxy = None;
        if let Some(cap) = regex::Regex::new(r#"HTTP_PROXY=(.+)"#).unwrap().captures(&content) {
            proxy = Some(cap[1].to_string());
        }
        let mut mirror = None;
        let daemon = session.read_file(DAEMON_JSON);
        if !daemon.is_empty() {
            if let Ok(cfg) = serde_json::from_str::<serde_json::Value>(&daemon) {
                if let Some(mirrors) = cfg.get("registry-mirrors").and_then(|v| v.as_array()) {
                    if let Some(first) = mirrors.first().and_then(|v| v.as_str()) {
                        mirror = Some(first.to_string());
                    }
                }
            }
        }
        ProxyStatus {
            component: ComponentId::DockerRemote,
            installed: self.detect(session),
            enabled,
            current_http_proxy: proxy.clone(),
            current_https_proxy: proxy,
            current_no_proxy: None,
            current_mirror: mirror,
            config_files: self.config_files(),
            manual_setup_steps: self
                .manual_steps()
                .into_iter()
                .map(|(t, c)| ManualStep {
                    title: t,
                    commands: c,
                })
                .collect(),
        }
    }
    fn enable(&self, session: &SshSession, config: &ProxyConfig) -> OpResult {
        if !session.has_sudo() {
            return OpResult::err("Sudo access required.");
        }
        let content = build_proxy_conf(config);
        if let Err(e) = session.write_file(DOCKER_CONF, &content, true) {
            return OpResult::err(format!("Failed to write {}: {}", DOCKER_CONF, e));
        }
        if !config.mirror.is_empty() {
            update_daemon_mirror(session, &config.mirror);
        }
        restart_docker(session);
        OpResult::ok("Docker proxy configured and restarted")
    }
    fn disable(&self, session: &SshSession) -> OpResult {
        if !session.has_sudo() {
            return OpResult::err("Sudo access required.");
        }
        session.run(&format!("sudo rm -f {}", DOCKER_CONF), 10);
        update_daemon_mirror(session, "");
        restart_docker(session);
        OpResult::ok("Docker proxy disabled")
    }
}

fn build_proxy_conf(config: &ProxyConfig) -> String {
    let mut lines = vec!["# Managed by proxy-switch".into(), "[Service]".into()];
    if !config.http_proxy.is_empty() {
        lines.push(format!(
            "Environment=\"HTTP_PROXY={}\"",
            config.http_proxy
        ));
        lines.push(format!(
            "Environment=\"http_proxy={}\"",
            config.http_proxy
        ));
    }
    if !config.https_proxy.is_empty() {
        lines.push(format!(
            "Environment=\"HTTPS_PROXY={}\"",
            config.https_proxy
        ));
        lines.push(format!(
            "Environment=\"https_proxy={}\"",
            config.https_proxy
        ));
    }
    if !config.no_proxy.is_empty() {
        lines.push(format!(
            "Environment=\"NO_PROXY={}\"",
            config.no_proxy
        ));
        lines.push(format!("Environment=\"no_proxy={}\"", config.no_proxy));
    }
    lines.push(String::new());
    lines.join("\n")
}

fn update_daemon_mirror(session: &SshSession, mirror: &str) {
    let existing = session.read_file(DAEMON_JSON);
    let mut cfg: serde_json::Value = if existing.trim().is_empty() {
        serde_json::Value::Object(serde_json::Map::new())
    } else {
        serde_json::from_str(&existing).unwrap_or(serde_json::Value::Object(serde_json::Map::new()))
    };
    if mirror.is_empty() {
        if let Some(obj) = cfg.as_object_mut() {
            obj.remove("registry-mirrors");
        }
    } else {
        if let Some(obj) = cfg.as_object_mut() {
            obj.insert(
                "registry-mirrors".into(),
                serde_json::Value::Array(vec![serde_json::Value::String(mirror.into())]),
            );
        }
    }
    let new_content = serde_json::to_string_pretty(&cfg).unwrap_or_default();
    session.write_file(DAEMON_JSON, &new_content, true).ok();
}

fn restart_docker(session: &SshSession) {
    session.run("sudo systemctl daemon-reload 2>&1", 15);
    session.run("sudo systemctl restart docker 2>&1", 30);
}
